packages/flutter/test/ui/window_test.dart

```diff
@@ -0,0 +1,183 @@
+// Copyright 2017 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:typed_data';
+import 'dart:ui';
+
+import 'package:test/test.dart';
+
+void main() {
+  group('window callback zones', () {
+    VoidCallback originalOnMetricsChanged;
+    VoidCallback originalOnLocaleChanged;
+    FrameCallback originalOnBeginFrame;
+    VoidCallback originalOnDrawFrame;
+    PointerDataPacketCallback originalOnPointerDataPacket;
+    VoidCallback originalOnSemanticsEnabledChanged;
+    SemanticsActionCallback originalOnSemanticsAction;
+    PlatformMessageCallback originalOnPlatformMessage;
+
+    setUp(() {
+      originalOnMetricsChanged = window.onMetricsChanged;
+      originalOnLocaleChanged = window.onLocaleChanged;
+      originalOnBeginFrame = window.onBeginFrame;
+      originalOnDrawFrame = window.onDrawFrame;
+      originalOnPointerDataPacket = window.onPointerDataPacket;
+      originalOnSemanticsEnabledChanged = window.onSemanticsEnabledChanged;
+      originalOnSemanticsAction = window.onSemanticsAction;
+      originalOnPlatformMessage = window.onPlatformMessage;
+    });
+
+    tearDown(() {
+      window.onMetricsChanged = originalOnMetricsChanged;
+      window.onLocaleChanged = originalOnLocaleChanged;
+      window.onBeginFrame = originalOnBeginFrame;
+      window.onDrawFrame = originalOnDrawFrame;
+      window.onPointerDataPacket = originalOnPointerDataPacket;
+      window.onSemanticsEnabledChanged = originalOnSemanticsEnabledChanged;
+      window.onSemanticsAction = originalOnSemanticsAction;
+      window.onPlatformMessage = originalOnPlatformMessage;
+    });
+
+    test('onMetricsChanged preserves callback zone', () {
```

Comment:
we should probably also have a test that verifies that the very first assignment to onBeginFrame (say) also correctly maintains the zone.

also we should verify that we can catch exceptions using this mechanism, since that's the whole point. :-) Ideally verifying that it works even when the callback is triggered by the engine itself... I think in tests we fire onBeginFrame at least once after you call scheduleFrame, maybe? Not sure if you can cause any of the others to trigger in a test.